js/src/main/scala-2.10/scalajson.ast/JValue.scala

```diff
@@ -143,9 +142,85 @@ final class JNumber private[ast] (val value: String) extends JValue {
 
   def copy(value: String): JNumber =
     value match {
-      case jNumberRegex(_ *) => new JNumber(value)
+      case jNumberRegex(_ *) => new JNumber(value)(0)
       case _ => throw new NumberFormatException(value)
     }
+
+  def toInt: Option[Long] = {
+    if ((constructedFlag & NumberFlags.int) == NumberFlags.int)
+      Some(value.toInt)
+    else {
+      try {
+        val asInt = value.toInt
+        if (BigInt(value) == BigInt(asInt))
+          Some(asInt)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toLong: Option[Long] = {
+    if ((constructedFlag & NumberFlags.long) == NumberFlags.long)
+      Some(value.toLong)
+    else {
+      try {
+        val asLong = value.toLong
+        if (BigInt(value) == BigInt(asLong))
+          Some(asLong)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigInt: Option[BigInt] = {
+    if ((constructedFlag & NumberFlags.bigInt) == NumberFlags.bigInt)
+      Some(BigInt(value))
+    else {
+      try {
+        Some(BigInt(value))
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigDecimal: Option[BigDecimal] = {
+    try {
+      Some(BigDecimal(value))
+    } catch {
+      case _: NumberFormatException => None
+    }
+  }
+
+  def toFloat: Option[Float] = {
+    if ((constructedFlag & NumberFlags.float) == NumberFlags.float)
+      Some(value.toFloat)
+    else {
+      val asFloat = value.toFloat
+      if (BigDecimal(value) == BigDecimal(asFloat.toDouble))
```

Comment:
TBH, I wasn't entirely sure how floats worked in Scala, but after reading this I think I need to do some more research If I have a better way of detecting losss of precision

To answer your questions
1. You can always read the raw the string value in a `JNumber`. The point of these converters is to correctly convert the string to a number (correctly here means without losing any data). As we can tell, this is quite hard ;)
2. Yes, else these converters are pointless. If its not possible then I will remove toFloat (nothing is stopping you from doing a `.toFloat` on the actual string value)
3. Well it currently returns `Option[Float]`. Returning `None` implies that you have lost some data (unless I am missing something here?)